### PR TITLE
gh-95882: fix regression in the traceback of exceptions propagated from inside a contextlib context manager

### PR DESCRIPTION
**Pull request by graingert**
…lib.asynccontextmanager

* Issue: gh-95882

### CODE REVIEW

**Comment by iritkatriel** [comment]
Some tests are failing.

**Comment by graingert** [comment]
Oh dear

File: Lib/contextlib.py [hunk 1]
```diff
@@ -239,6 +240,7 @@ async def __aexit__(self, typ, value, traceback):
                     isinstance(value, (StopIteration, StopAsyncIteration))
                     and exc.__cause__ is value
                 ):
+                    exc.__traceback__ = traceback
```
Comment (graingert) [inline Lib/contextlib.py]: needs a test for this case

Comment (graingert) [reply Lib/contextlib.py]: @iritkatriel it doesn't look like this line does anything. If I delete it I get the same behavior, same with L176:

https://github.com/python/cpython/blob/adf2bf3cc554b7c841edce47bb023bbc89d50908/Lib/contextlib.py#L166-L177

File: Lib/test/test_contextlib.py [hunk 2]
```diff
@@ -104,15 +104,41 @@ def f():
...
+                        raise stop_exc
+                except type(stop_exc) as e:
+                    self.assertIs(e, stop_exc)
+                    frames = traceback.extract_tb(e.__traceback__)
+                else:
+                    self.fail(f'{stop_exc} was suppressed')
+
+                self.assertEqual(len(frames), 2)
+                self.assertEqual(frames[0].name, 'f')
+                self.assertEqual(frames[0].line, 'yield')
```
Comment (graingert) [inline Lib/test/test_contextlib.py]: @iritkatriel do you expect to see this `yield` here?

Comment (iritkatriel) [reply Lib/test/test_contextlib.py]: What does 3.10 do?

Comment (graingert) [reply Lib/test/test_contextlib.py]: the `yield` is new, incorrect, behavior in 3.11:


```python
import contextlib
import logging

logger = logging.getLogger("__name__")

@contextlib.contextmanager
def f():
    yield


try:
    with f():
        raise StopIteration
except StopIteration:
    logger.exception("stop iteration")



print("================ overwriting StopIteration via globals!==============")


class StopIteration(Exception):
    pass


try:
    with f():
        raise StopIteration
except StopIteration:
    logger.exception("stop iteration")
```


```
 graingert@conscientious  testing311  ~/projects/trio   master ± python3.10 demo.py
stop iteration
Traceback (most recent call last):
  File "/home/graingert/projects/trio/demo.py", line 13, in <module>
    raise StopIteration
StopIteration
================ overwriting StopIteration via globals!==============
stop iteration
Traceback (most recent call last):
  File "/home/graingert/projects/trio/demo.py", line 28, in <module>
    raise StopIteration
StopIteration
```

Notice the extra `yield` in the first run before `overwriting StopIteration via globals` happens:

```
 graingert@conscientious  testing311  ~/projects/trio   master ± python3.11 demo.py 
stop iteration
Traceback (most recent call last):
  File "/home/graingert/projects/trio/demo.py", line 8, in f
    yield
  File "/home/graingert/projects/trio/demo.py", line 13, in <module>
    raise StopIteration
StopIteration
================ overwriting StopIteration via globals!==============
stop iteration
Traceback (most recent call last):
  File "/home/graingert/projects/trio/demo.py", line 28, in <module>
    raise StopIteration
StopIteration
```

File: Lib/contextlib.py (hunk 1 above)
Comment (graingert) [inline Lib/contextlib.py]:
```suggestion
                    value.__traceback__ = traceback
```

File: Lib/test/test_contextlib.py (hunk 2 above)
Comment (graingert) [reply Lib/test/test_contextlib.py]: fixed it!

**Comment by iritkatriel** [comment]
Could you make a PR against 3.10 to backport only the new tests, so we can make sure this is not changing anything?

File: Misc/NEWS.d/next/Library/2022-08-11-10-02-19.gh-issue-95882.FsUr72.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Fix a 3.11 regression in :func:`~contextlib.asynccontextmanager`, which caused it to propagate exceptions with incorrect tracebacks.
```
Comment (graingert) [inline Misc/NEWS.d/next/Library/2022-08-11-10-02-19.gh-issue-95882.FsUr72.rst]:
```suggestion
Fix a 3.11 regression in :func:`~contextlib.asynccontextmanager`, which caused it to propagate exceptions with incorrect tracebacks and fix a 3.11 regression in  :func:`~contextlib.asynccontextmanager`, which caused it to propagate exceptions with incorrect tracebacks for :exc:`StopIteration`
```

File: Misc/NEWS.d/next/Library/2022-08-11-10-02-19.gh-issue-95882.FsUr72.rst [hunk 4]
```diff
@@ -0,0 +1 @@
+Fix a 3.11 regression in :func:`~contextlib.asynccontextmanager`, which caused it to propagate exceptions with incorrect tracebacks and fix a 3.11 regression in  :func:`~contextlib.asynccontextmanager`, which caused it to propagate exceptions with incorrect tracebacks for :exc:`StopIteration`
```
Comment (graingert) [inline Misc/NEWS.d/next/Library/2022-08-11-10-02-19.gh-issue-95882.FsUr72.rst]:
```suggestion
Fix a 3.11 regression in :func:`~contextlib.asynccontextmanager`, which caused it to propagate exceptions with incorrect tracebacks and fix a 3.11 regression in  :func:`~contextlib.contextmanager`, which caused it to propagate exceptions with incorrect tracebacks for :exc:`StopIteration`.
```

Comment (graingert) [reply Misc/NEWS.d/next/Library/2022-08-11-10-02-19.gh-issue-95882.FsUr72.rst]: > Could you make a PR against 3.10 to backport only the new tests, so we can make sure this is not changing anything?

hopefully I can get to see what happens all the way to 3.7 too

File: Misc/NEWS.d/next/Library/2022-08-11-10-02-19.gh-issue-95882.FsUr72.rst [hunk 5]
```diff
@@ -0,0 +1 @@
+Fix a 3.11 regression in :func:`~contextlib.asynccontextmanager`, which caused it to propagate exceptions with incorrect tracebacks and fix a 3.11 regression in  :func:`~contextlib.contextmanager`, which caused it to propagate exceptions with incorrect tracebacks for :exc:`StopIteration`.
```
Comment (graingert) [inline Misc/NEWS.d/next/Library/2022-08-11-10-02-19.gh-issue-95882.FsUr72.rst:1]:
```suggestion
Fix a 3.11 regression in :func:`~contextlib.asynccontextmanager`, which caused it to propagate exceptions with incorrect tracebacks and fix a 3.11 regression in  :func:`~contextlib.contextmanager`, which caused it to propagate exceptions with incorrect tracebacks for :exc:`StopIteration`.  Patch by Irit Katriel adapted for async by Thomas Grainger.
```

**Comment by graingert** [comment]
> Could you make a PR against 3.10 to backport only the new tests, so we can make sure this is not changing anything?

Yeah, I'll see how far I can backport it in contextlib2 too

**Comment by graingert** [comment]
@iritkatriel can you re-review this please?

**Comment by iritkatriel** [comment]
Is [this](https://github.com/python/cpython/pull/95883#issuecomment-1212228959) done?

**Comment by graingert** [comment]
Once it's in a release I'll apply it to the backport

**Comment by iritkatriel** [comment]
> Once it's in a release I'll apply it to the backport

It has to be a separate PR anyway because it will only contain the tests. Why not do it as part of the review process for this PR?

**Review by iritkatriel — APPROVED** [review]

**Comment by iritkatriel** [comment]
This looks fine but I think we should see the tests pass on all 3.10 buildbots before merging this. It’s the same amount of work, we might as well do it in the right order.

**Pull request merged**